### PR TITLE
Added validation for identical OPM rate descriptions - FFY 2023 measures.

### PR DESCRIPTION
**Pull request by keeysnc**
### Description
Create a validation that fires when states create data labels for rates in the Other Performance Section that are identical (e.g. two rates that are both labeled 18-64 with no other distinction). These labels must be distinct from one another to be read into the Mathematica database correctly.

When fields match and user validates measure display error message:

H1: Other Performance Measure Error
Body: Measure description must be unique 

* The validation only notifies user. They can still complete the form.

### Related ticket(s)
https://qmacbis.atlassian.net/jira/software/c/projects/MDCT/boards/232?modal=detail&selectedIssue=MDCT-2333
MDCT-2333

https://user-images.githubusercontent.com/14514294/235723363-61986e9e-41b6-4f5e-99f0-cbac40b9e2c5.mov

---
### How to test
1. Login as user
2. Go to 2023 FY
3. Select a core set
4. Select a measure
5. In "Measurement Specification" select 'Other'
6. Scroll down to Other Performance Measure and add two performance measures with identical descriptions in the 'Describe the Rate' section.
7. Validate
8. You should see an 'Other Performance Measure Error' validation message below

- [x] I have performed a self-review of my code
- [x] I have added [thorough](https://bit.ly/3zPrxuZ) tests, if necessary
- [ ] I have updated relevant documentation, if necessary
---

_convert to a different template: [test → val](?expand=1&template=test-to-val-deployment.md)_ | _[val → prod](?expand=1&template=val-to-prod-deployment.md)_

### CODE REVIEW

**Review by tblackwe — DISMISSED** [review]
LGTM

**Review by ailZhou — COMMENTED** [review]
How smart does this need to be? It works if the text is exact but if there's leading or other spaces, it will bypass the error.  ex. '(space)18 to 64' to '18 to 64'. 

Not sure if intentional. Otherwise, looks and works great.

**Comment by keeysnc** [comment]
> How smart does this need to be? It works if the text is exact but if there's leading or other spaces, it will bypass the error. ex. '(space)18 to 64' to '18 to 64'.
> 
> Not sure if intentional. Otherwise, looks and works great.

@ailZhou Good point, I can look into that to see if there's already a trim but if not I can try and add one in.

File: services/ui-src/src/measures/2023/shared/globalValidations/validateOPMRates/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  it("Should throw error if there are duplicate opm rates", () => {
+    formData[DC.OPM_RATES][0][DC.DESCRIPTION] = "18-64";
+    formData[DC.OPM_RATES][1][DC.DESCRIPTION] = "18-64";
+    const errorMessage = "Measure description must be unique.";
+    const errorArray = run_validation(formData);
+    expect(errorArray.length).toBe(1);
+    expect(errorArray[0].errorMessage).toBe(errorMessage);
+  });
+});
```
Comment (benmartin-coforma) [inline services/ui-src/src/measures/2023/shared/globalValidations/validateOPMRates/index.test.ts:25]: Small suggestion here, to increase test coverage.

```suggestion

  it("Should not throw error if there are no duplicates", () => {
    formData[DC.OPM_RATES][0][DC.DESCRIPTION] = "18-30";
    formData[DC.OPM_RATES][1][DC.DESCRIPTION] = "31-64";
    const errorArray = run_validation(formData);
    expect(errorArray.length).toBe(0);
  });

  it("Should not treat multiple empty descriptions as duplicates", () => {
    formData[DC.OPM_RATES][0][DC.DESCRIPTION] = "";
    formData[DC.OPM_RATES][1][DC.DESCRIPTION] = "";
    const errorArray = run_validation(formData);
    expect(errorArray.length).toBe(0);
  });
});
```

**Review by benmartin-coforma — APPROVED** [review]
Looks pretty good!

File: services/ui-src/src/measures/2023/shared/globalValidations/validateOPMRates/index.test.ts (hunk 1 above)
Comment (keeysnc) [reply services/ui-src/src/measures/2023/shared/globalValidations/validateOPMRates/index.test.ts:30]: Updated. Thanks!

File: services/ui-src/src/measures/2023/shared/globalValidations/validateOPMRates/index.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+export const validateOPMRates = (
+  otherPerformanceMeasure: OtherPerformanceMeasure["OtherPerformanceMeasure-Rates"],
+  errorMessage?: string
+) => {
+  const errorArray: FormError[] = [];
+
+  if (otherPerformanceMeasure) {
+    const opm_descriptions = otherPerformanceMeasure.filter(
+      (item: any) => item.description !== ""
```
Comment (ailZhou) [inline services/ui-src/src/measures/2023/shared/globalValidations/validateOPMRates/index.ts:14]:
```suggestion
      (item: any) => !!item.description
```
slight code recommendation to check for undefined values as well, i had ran into errors validating before i wrote anything into the text field

**Review by ailZhou — COMMENTED** [review]
Question: i was able to submit two description rate as empty string with no validation warning, is that desired?

**Comment by keeysnc** [comment]
> Question: i was able to submit two description rate as empty string with no validation warning, is that desired?

@ailZhou After talking with David it seems like its the desired functionality.

**Review by ailZhou — APPROVED** [review]
Looks good to me!

**Review by benmartin-coforma — APPROVED** [review]

**Review by karla-vm — APPROVED** [review]
Awesome! 💯

**Pull request merged**